File: client-dotnet-httpclient/src/Epistola.Client/Error/ProblemDetailException.cs
using System.Collections.Generic;
using System.Net;
using Epistola.Client.Client;
using Epistola.Client.Model;

namespace Epistola.Client.Error;

/// <summary>
/// An <see cref="ApiException"/> carrying a parsed
/// <a href="https://www.rfc-editor.org/rfc/rfc9457">RFC 9457</a> <see cref="ProblemDetail"/>
/// (<c>application/problem+json</c>) body.
///
/// Thrown by the opt-in <see cref="ProblemDetailHandler"/>. It extends the generated
/// <see cref="ApiException"/> on purpose: existing <c>catch (ApiException)</c> sites keep working
/// and consumers retain the inherited <see cref="ApiException.ErrorCode"/> /
/// <see cref="ApiException.Headers"/> accessors.
///
/// The machine-readable discriminator is the problem <see cref="Type"/> URI; switch on
/// <see cref="TypeSlug"/>. Field-level validation errors (the <c>ValidationProblemDetail</c> shape)
/// are surfaced via <see cref="Errors"/>; per-example data-model validation failures (the
/// <c>DataModelValidationProblemDetail</c> shape, <c>data-model-validation-error</c>) via
/// <see cref="ValidationErrors"/> — the generated <c>ProblemDetail</c>,
/// <c>ValidationProblemDetail</c>, and <c>DataModelValidationProblemDetail</c> models are
/// independent, so the base fields and each extension are carried separately.
/// </summary>
public class ProblemDetailException : ApiException
{
    /// <summary>The parsed base problem (<c>type</c>, <c>title</c>, <c>status</c>, <c>detail</c>, <c>instance</c>).</summary>
    public ProblemDetail Problem { get; }

    /// <summary>Field-level validation errors when the body was a <c>ValidationProblemDetail</c>, else empty.</summary>
    public IReadOnlyList<ValidationError> Errors { get; }

    /// <summary>
    /// Per-example data-model validation failures (example name → failures) when the body was a
    /// <c>DataModelValidationProblemDetail</c> (<c>data-model-validation-error</c>, 422), else empty.
    /// </summary>
    public IReadOnlyDictionary<string, List<DataModelValidationError>> ValidationErrors { get; }

    /// <summary>The HTTP status of the error response.</summary>
    public HttpStatusCode StatusCode { get; }

    public ProblemDetailException(
        ProblemDetail problem,
        IReadOnlyList<ValidationError> errors,
        IReadOnlyDictionary<string, List<DataModelValidationError>> validationErrors,
        HttpStatusCode statusCode,
        string? rawBody,
        Multimap<string, string>? headers)
        : base((int)statusCode, BuildMessage(statusCode, problem), rawBody, headers)
    {
        Problem = problem;
        Errors = errors;
        ValidationErrors = validationErrors;
        StatusCode = statusCode;
    }

    /// <summary>The problem <c>type</c> URI (<c>about:blank</c> when unspecified).</summary>
    public string Type => Problem.Type ?? ProblemTypes.BlankType;

    /// <summary>
    /// Kebab-case slug derived from <see cref="Type"/> by stripping <see cref="ProblemTypes.TypeBase"/>,
    /// or <c>null</c> for <c>about:blank</c> and non-Epistola types. Compare against <c>KnownProblemSlugs</c>.
    /// </summary>
    public string? TypeSlug => ProblemTypes.SlugFor(Problem.Type);

    /// <summary>Short human-readable summary of the problem type (RFC 9457 <c>title</c>).</summary>
    public string? Title => Problem.Title;

    /// <summary>The HTTP status carried in the problem body (RFC 9457 <c>status</c>).</summary>
    public int ProblemStatus => Problem.Status;

    /// <summary>Occurrence-specific explanation (RFC 9457 <c>detail</c>), if the server provided one.</summary>
    public string? Detail => Problem.Detail;

    /// <summary>True when this problem carried field-level validation errors.</summary>
    public bool IsValidationProblem => Errors.Count > 0;

    /// <summary>True when this problem carried per-example data-model validation failures.</summary>
    public bool IsDataModelValidationProblem => ValidationErrors.Count > 0;

    private static string BuildMessage(HttpStatusCode status, ProblemDetail problem)
    {
        var title = problem.Title ?? status.ToString();
        return string.IsNullOrEmpty(problem.Detail)
            ? $"{(int)status} {title}"
            : $"{(int)status} {title}: {problem.Detail}";
    }
}

File: contracts/api/clients/kotlin-spring-restclient/src/main/kotlin/app/epistola/client/error/ProblemDetailException.kt
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

package app.epistola.client.error

import app.epistola.client.model.DataModelValidationError
import app.epistola.client.model.ProblemDetail
import app.epistola.client.model.ValidationError
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatusCode
import org.springframework.web.client.RestClientResponseException
import java.net.URI
import java.nio.charset.Charset

/**
 * A [RestClientResponseException] carrying a parsed [RFC 9457](https://www.rfc-editor.org/rfc/rfc9457)
 * [ProblemDetail] (`application/problem+json`) body.
 *
 * Thrown by the opt-in error handler installed via [installProblemDetailHandler]. It extends
 * [RestClientResponseException] on purpose: the generated APIs declare
 * `@Throws(RestClientResponseException::class)`, so existing `catch (e: RestClientResponseException)`
 * sites keep working and consumers retain the inherited `statusCode` /
 * [getResponseBodyAsString] accessors. (Note: because it is a sibling of `HttpClientErrorException`
 * / `HttpServerErrorException`, it is *not* caught by `catch (e: HttpClientErrorException)` — catch
 * [ProblemDetailException] or [RestClientResponseException].)
 *
 * The machine-readable discriminator is the problem [type] URI; switch on [typeSlug]. Field-level
 * validation errors (the `ValidationProblemDetail` shape from the contract) are surfaced via
 * [errors]; per-example data-model validation failures (the `DataModelValidationProblemDetail`
 * shape, `data-model-validation-error`) are surfaced via [validationErrors] — `ProblemDetail`,
 * `ValidationProblemDetail`, and `DataModelValidationProblemDetail` are independent classes, so the
 * base fields and each extension are carried separately. Any *other* member a problem body carries —
 * one the contract does not name, such as `catalog-schema-too-old`'s `version` / `baselineVersion` —
 * is in [extensions], since [ProblemDetail] is hand-written with a catch-all rather than generated
 * as a closed set of fields.
 */
class ProblemDetailException(
    /** The parsed base problem (`type`, `title`, `status`, `detail`, `instance`). */
    val problem: ProblemDetail,
    /** Field-level validation errors when the body was a `ValidationProblemDetail`, else empty. */
    val errors: List<ValidationError>,
    /**
     * Per-example data-model validation failures (example name → failures) when the body was a
     * `DataModelValidationProblemDetail` (`data-model-validation-error`, 422), else empty.
     */
    val validationErrors: Map<String, List<DataModelValidationError>> = emptyMap(),
    statusCode: HttpStatusCode,
    statusText: String,
    headers: HttpHeaders?,
    responseBody: ByteArray?,
    responseCharset: Charset?,
) : RestClientResponseException(
    buildMessage(statusCode, problem),
    statusCode,
    statusText,
    headers,
    responseBody,
    responseCharset,
) {
    /** The problem `type` URI (`about:blank` when unspecified). */
    val type: URI get() = problem.type

    /**
     * Kebab-case slug derived from [type] by stripping [ProblemTypes.TYPE_BASE], or `null` for
     * `about:blank` and non-Epistola types. Compare against [KnownProblemSlugs].
     */
    val typeSlug: String? get() = ProblemTypes.slugFor(problem.type)

    /** Short human-readable summary of the problem type (RFC 9457 `title`). */
    val title: String get() = problem.title

    /**
     * The HTTP status carried in the problem body. Usually equal to the response
     * [getStatusCode], but named distinctly to avoid clashing with the inherited member.
     */
    val problemStatus: Int get() = problem.status

    /** Occurrence-specific explanation (RFC 9457 `detail`), if the server provided one. */
    val detail: String? get() = problem.detail

    /**
     * Every problem member outside `type`/`title`/`status`/`detail`/`instance`, keyed by its JSON
     * name — `catalog-schema-too-old`'s `version` and `baselineVersion`, or any extension member on
     * any problem type the contract adds later, without needing a client release to read it.
     */
    val extensions: Map<String, Any?> get() = problem.extensions

    /** True when this problem carried field-level validation errors. */
    val isValidationProblem: Boolean get() = errors.isNotEmpty()

    /** True when this problem carried per-example data-model validation failures. */
    val isDataModelValidationProblem: Boolean get() = validationErrors.isNotEmpty()

    private companion object {
        fun buildMessage(status: HttpStatusCode, problem: ProblemDetail): String = "$status ${problem.title}" + (problem.detail?.let { ": $it" } ?: "")
    }
}
